File: day01/main.scala
object Day01 {
  def fuel_needed(mass: Int): Int = {
    math.max((mass / 3) - 2, 0)
  }

  def total_fuel_needed(mass: Int): Int = {
    val value = fuel_needed(mass)

    if (value <= 0)
      0
    else
      total_fuel_needed(value) + value
  }

  def main(args: Array[String]): Unit = {
    val lines = scala.io.Source.fromFile("input").mkString
    val modules = lines.split("\n").toSeq.map(_.trim).filter(_ != "").map(_.toInt)

    val initial_fuel = modules.map(x => fuel_needed(x)).sum
    val total_fuel = modules.map(x => total_fuel_needed(x)).sum

    println(initial_fuel)
    println(total_fuel)
  }
}

File: day01/main.go
package main

import (
	"io/ioutil"
	"fmt"
	"strings"
	"strconv"
)

func parseInput(data string) []int {
	input := strings.Trim(data, "\t \n")
	lines := strings.Split(input, "\n")

	modules := []int{}

	for _, i := range lines {
		j, err := strconv.Atoi(i)
		if err != nil {
			panic(err)
		}

		modules = append(modules, j)
	}

	return modules
}

func fuel_needed(mass int) int {
	fuel := (mass / 3) - 2

	if fuel <= 0 {
		return 0
	}

	return fuel
}

func initial_fuel_needed(modules []int) int {
	total := 0

	for _, mass := range modules {
		total += fuel_needed(mass)
	}

	return total
}

func total_fuel_needed(modules []int) int {
	total := 0

	for _, mass := range modules {
		fuel := fuel_needed(mass)

		for {
			if fuel <= 0 {
				break
			}
			total += fuel

			fuel = fuel_needed(fuel)
		}
	}

	return total
}

func main() {
	data, _ := ioutil.ReadFile("input")
	modules := parseInput(string(data))

	initial_fuel := initial_fuel_needed(modules)
	total_fuel := total_fuel_needed(modules)

	fmt.Println(initial_fuel)
	fmt.Println(total_fuel)
}
